waveguide_mode_converter/mode_converter_meep_opt.py

```diff
@@ -334,80 +416,88 @@ def J(refl_mon,tran_mon):
 
     # initial guess for design parameters
     x = np.ones((n,)) * 0.5
+    x[Si_mask.flatten()] = 1    # set the edges of waveguides to silicon
+    x[SiO2_mask.flatten()] = 0  # set the other edges to SiO2
 
     # lower and upper bounds for design weights
     lb = np.zeros((n,))
+    lb[Si_mask.flatten()] = 1
     ub = np.ones((n,))
+    ub[SiO2_mask.flatten()] = 0
 
     # insert epigraph variable and bounds as first element of 1d arrays
-    x = np.insert(x, 0, 1.2)  # initial guess for the worst error (max: 2.0)
-    lb = np.insert(lb, 0, 0)  # lower bound: cannot be less than 0
-    ub = np.insert(ub, 0, 2)  # upper bound: cannot be more than 2
+    x = np.insert(x, 0, 1.2)   # initial guess for the worst error
+    lb = np.insert(lb, 0, 0.)  # lower bound: cannot be less than 0.
+    ub = np.insert(ub, 0, 2.)  # upper bound: cannot be more than 2.
 
     evaluation_history = []
     cur_iter = [0]
 
-    cur_beta = 4
-    beta_scale = 2
-    num_betas = 8
-    max_eval = 12
+    betas = [8, 16, 32]
+    max_eval = 80
     tol = np.array([1e-6] * opt.nf)
-    for iters in range(num_betas):
+    for beta in betas:
         solver = nlopt.opt(algorithm, n + 1)
         solver.set_lower_bounds(lb)
         solver.set_upper_bounds(ub)
         solver.set_min_objective(f)
         solver.set_maxeval(max_eval)
         solver.add_inequality_mconstraint(
-            lambda r, x, g: c(r, x, g, eta_i, cur_beta),
+            lambda r, x, g: c(r, x, g, eta_i, beta),
             tol,
         )
+        # apply the minimum linewidth constraint
+        # only in the final "epoch"
+        if beta == betas[-1]:
+            solver.add_inequality_mconstraint(
+                lambda r, x, g: glc(r, x, g, beta),
+                [1e-8] * opt.nf,
+            )
         x[:] = solver.optimize(x)
```

Comment:
So we really need to recalibrate the dummy parameter (`t`) before starting the optimization. You can see the optimizer is spending most of its time adjusting this during the second epoch.

You can do one of three things (in order of both increasing difficulty and increased expected performance):
1. Reset `t` to the same default at the start of each epoch. Heuristically, it's much easier for the optimizer to bring it down (and all the constraints are satisfied) than to bring it up (where none of the constraints are satisfied).
2. Run a single forward run _before_ starting each new optimization epoch, and manually set `t` as the maximum value.
3. Modify the CCSA algorithm to update `t` analytically (as described in the original [Svanberg paper](https://epubs.siam.org/doi/10.1137/S1052623499362822)).